src/java.base/share/classes/jdk/internal/classfile/impl/AnnotationReader.java

```diff
@@ -0,0 +1,282 @@
+        for (int i = 0; i < numAnnotations; ++i) {
+            annos[i] = readAnnotation(classReader, pos);
+            pos = skipAnnotation(classReader, pos);
+        }
+        return SharedSecrets.getJavaUtilCollectionAccess().listFromTrustedArrayNullsAllowed(annos);
+    }
+
+    public static AnnotationValue readElementValue(ClassReader classReader, int p) {
+        char tag = (char) classReader.readU1(p);
+        ++p;
+        return switch (tag) {
+            case 'B' -> new AnnotationImpl.OfByteImpl((IntegerEntry)classReader.readEntry(p));
+            case 'C' -> new AnnotationImpl.OfCharacterImpl((IntegerEntry)classReader.readEntry(p));
+            case 'D' -> new AnnotationImpl.OfDoubleImpl((DoubleEntry)classReader.readEntry(p));
+            case 'F' -> new AnnotationImpl.OfFloatImpl((FloatEntry)classReader.readEntry(p));
+            case 'I' -> new AnnotationImpl.OfIntegerImpl((IntegerEntry)classReader.readEntry(p));
+            case 'J' -> new AnnotationImpl.OfLongImpl((LongEntry)classReader.readEntry(p));
+            case 'S' -> new AnnotationImpl.OfShortImpl((IntegerEntry)classReader.readEntry(p));
+            case 'Z' -> new AnnotationImpl.OfBooleanImpl((IntegerEntry)classReader.readEntry(p));
+            case 's' -> new AnnotationImpl.OfStringImpl(classReader.readUtf8Entry(p));
+            case 'e' -> new AnnotationImpl.OfEnumImpl(classReader.readUtf8Entry(p), classReader.readUtf8Entry(p + 2));
+            case 'c' -> new AnnotationImpl.OfClassImpl(classReader.readUtf8Entry(p));
+            case '@' -> new AnnotationImpl.OfAnnotationImpl(readAnnotation(classReader, p));
+            case '[' -> {
+                int numValues = classReader.readU2(p);
+                p += 2;
+                var values = new Object[numValues];
+                for (int i = 0; i < numValues; ++i) {
+                    values[i] = readElementValue(classReader, p);
+                    p = skipElementValue(classReader, p);
+                }
+                yield new AnnotationImpl.OfArrayImpl(SharedSecrets.getJavaUtilCollectionAccess().listFromTrustedArrayNullsAllowed(values));
+            }
+            default -> throw new IllegalArgumentException(
+                    "Unexpected tag '%s' in AnnotationValue, pos = %d".formatted(tag, p - 1));
+        };
+    }
+
+    public static List<TypeAnnotation> readTypeAnnotations(ClassReader classReader, int p, LabelContext lc) {
+        int numTypeAnnotations = classReader.readU2(p);
+        p += 2;
+        var annotations = new Object[numTypeAnnotations];
+        for (int i = 0; i < numTypeAnnotations; ++i) {
+            annotations[i] = readTypeAnnotation(classReader, p, lc);
+            p = skipTypeAnnotation(classReader, p);
+        }
+        return SharedSecrets.getJavaUtilCollectionAccess().listFromTrustedArrayNullsAllowed(annotations);
+    }
+
+    public static List<List<Annotation>> readParameterAnnotations(ClassReader classReader, int p, boolean isVisible) {
```

Comment:
Parameter `isVisible` is unused, but is called with true/false values.

Reply:
fixed, thanks.